File: library/kotlin/src/io/envoyproxy/envoymobile/stats/HistogramImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import java.lang.ref.WeakReference

/**
 * Envoy implementation of a `Histogram` for measurements of generic int values
 */
internal class HistogramImpl : Histogram {
  internal val envoyEngine: WeakReference<EnvoyEngine>
  internal val elements: String

  internal constructor(engine: EnvoyEngine, elements: List<Element>) {
    this.envoyEngine = WeakReference<EnvoyEngine>(engine)
    this.elements = elements.joinToString(separator = ".") { it.value }
  }

  override fun recordValue(value: Int) {
    envoyEngine.get()?.recordHistogramValue(elements, value)
  }
}

File: library/swift/src/stats/HistogramImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// The implementation of time series gauge.
@objcMembers
class HistogramImpl: NSObject, Gauge {
  private let series: String
  private weak var engine: EnvoyEngine?

  init(elements: [Element], engine: EnvoyEngine) {
    self.series = elements.map { $0.value }.joined(separator: ".")
    self.engine = engine
    super.init()
  }

  /// Record a new value for the histogram distribution.
  /// TODO: potentially raise error to platform if the operation is not successful.
  func recordValue(value: Int) {
    guard let engine = self.engine else {
      return
    }

    engine.recordHistogram(self.series, value: numericCast(value))
  }
}
